### PR TITLE
feat: ScalarValue from String

### LINKED ISSUE #8380: Make it easier to create `ScalarValue` from `String` and `&str`
### Is your feature request related to a problem or challenge?

Someone reported on Discord: https://discord.com/channels/885562378132000778/1166447479609376850/1179899297483980941

>  currently when I'm trying to pass a String into ScalarValue::from(), the compiler is telling me that the trait bound ScalarValue: From<std::string::String> is not satisfied

Let's make it easier for people to get started with DataFusioN!

### Describe the solution you'd like

Make it so the following code works

```rust
let scalar = ScalarValue::from("foo");
let scalar = ScalarValue::from("foo".to_string());
```

### Describe alternatives you've considered

_No response_

### Additional context

I think this is a good first issue as it is well specified and would be a nice usability improvement

Related PR: https://github.com/apache/arrow-datafusion/pull/8378 (it would be nice to add an example of creating ScalarValue from a string there too)

### PR DESCRIPTION
**Pull request by QuenKar**
## Which issue does this PR close?

Closes #8380 .

### CODE REVIEW

**Review by Dandandan — APPROVED** [review]

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -4645,6 +4651,16 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_scalar_value_from_string() {
+        let scalar = ScalarValue::from("foo");
+        assert_eq!(scalar, ScalarValue::Utf8(Some("foo".to_string())));
+        let scalar = ScalarValue::from("foo".to_string());
+        assert_eq!(scalar, ScalarValue::Utf8(Some("foo".to_string())));
```
[the remark this answers is not in the document]
Comment (QuenKar) [reply datafusion/common/src/scalar.rs:4659]: I find that there is no code similar to `"foo".to_string().as_str() `or `some_string.as_str()` or `&some_string` or `as_str().into()` in this use(But maybe I don't find all of them...

[the remark this answers is not in the document]
Comment (Dandandan) [reply datafusion/common/src/scalar.rs:4659]: One tricky might be removing `From<&str>` implementqtion and seeing if something fails to compile.

**Comment by QuenKar** [comment]
nice idea！

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Dani?l Heres ***@***.***>
发送时间: Tuesday, December 5, 2023 6:35:21 PM
收件人: apache/arrow-datafusion ***@***.***>
抄送: Wei ***@***.***>; Mention ***@***.***>
主题: Re: [apache/arrow-datafusion] feat: ScalarValue from String (PR #8411)


@Dandandan commented on this pull request.

________________________________

In datafusion/common/src/scalar.rs<https://github.com/apache/arrow-datafusion/pull/8411#discussion_r1415328153>:

> +        let scalar = ScalarValue::from("foo".to_string());
+        assert_eq!(scalar, ScalarValue::Utf8(Some("foo".to_string())));


One tricky might be removing From<&str> implementqtion and seeing if something fails to compile.

―

File: datafusion/common/src/scalar.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/common/src/scalar.rs:4659]: > I think the only difference is From<String> avoids copying the string content?

It avoids copying as well as makes creating `ScalarValue`s easier to create / user (e.g. someone was confused as reported on https://github.com/apache/arrow-datafusion/issues/8380)

> One tricky might be removing From<&str> implementqtion and seeing if something fails to compile.

I tried this and found there weren't any obvious examples in the codebase -- the more common pattern is `ScalarValue::from("foo")` with a constant string

**Review by alamb — APPROVED** [review]
Thank you @QuenKar  and @Dandandan and @viirya

File: datafusion/common/src/scalar.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (viirya) [reply datafusion/common/src/scalar.rs:4659]: > I tried this and found there weren't any obvious examples in the codebase -- the more common pattern is `ScalarValue::from("foo")` with a constant string

Nice!

**Pull request merged**